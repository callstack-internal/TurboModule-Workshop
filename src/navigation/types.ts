import { PerroConEstado } from '../types/dog.ts';

export type RouteNamesParamList = RootStackParamList | DrawerParamList;

export type RootStackParamList = {
  Home: undefined;
  AdoptedDogs: undefined;
  DogDetails: { dog: PerroConEstado };
  HowItWorks: undefined;
};

export type DrawerParamList = {
  Main: undefined;
  HowItWorks: undefined;
};
