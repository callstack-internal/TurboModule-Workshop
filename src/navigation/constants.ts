export const Screens = {
  Main: {
    title: 'Catálogo de Perros',
    name: 'Main',
  },
  HowItWorks: {
    title: 'Cómo funciona',
    name: 'HowItWorks',
  },
  Home: {
    title: 'Home',
    name: 'Home',
  },
  DogDetails: {
    title: 'Detalles',
    name: 'DogDetails',
  },
  Adoptar: {
    title: 'Adoptar',
    name: 'Adoptar',
  },
  MisPerros: {
    title: 'Mis Perros',
    name: 'MisPerros',
  },
} as const;
