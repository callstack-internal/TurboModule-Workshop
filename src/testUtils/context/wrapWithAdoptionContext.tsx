import { AdoptionProvider } from '../../context/AdoptionContext.tsx';

export const wrapWithAdoptionContext = (children: React.ReactNode) => {
  return <AdoptionProvider>{children}</AdoptionProvider>;
};
