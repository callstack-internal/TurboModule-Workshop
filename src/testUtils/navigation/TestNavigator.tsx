import React, { useEffect } from 'react';

import {
  createNavigatorFactory,
  useNavigationBuilder,
  StackRouter,
  TabRouter,
  DrawerRouter,
  DefaultRouterOptions,
  StackRouterOptions,
  TabRouterOptions,
  DrawerRouterOptions,
  NavigationHelpers,
  ParamListBase,
  RouterFactory,
  Router,
  NavigationState,
  NavigationAction,
  NavigationContainer,
} from '@react-navigation/native';
import type {
  StackNavigationState,
  TabNavigationState,
  DrawerNavigationState,
} from '@react-navigation/native';
import type {
  StackActionHelpers,
  TabActionHelpers,
  DrawerActionHelpers,
} from '@react-navigation/native';
import { Text } from 'react-native';

type TestScreenOptions = Record<string, unknown>;

type NavigatorType = 'stack' | 'tabs' | 'drawer';

interface TestNavigatorImplProps extends DefaultRouterOptions {
  type: NavigatorType;
  screenOptions?: TestScreenOptions;
  initialRouteName?: string;
  children: React.ReactNode;
}

type AnyNavigationState =
  | StackNavigationState<ParamListBase>
  | TabNavigationState<ParamListBase>
  | DrawerNavigationState<ParamListBase>;

type AnyRouterOptions = StackRouterOptions | TabRouterOptions | DrawerRouterOptions;

type AnyActionHelpers =
  | StackActionHelpers<ParamListBase>
  | TabActionHelpers<ParamListBase>
  | DrawerActionHelpers<ParamListBase>;

type ExtractedAction = Exclude<
  Parameters<Router<NavigationState, NavigationAction>['getStateForAction']>[0],
  undefined
>;

/**
 * Custom navigator component used for testing React Navigation screens.
 * Supports `stack`, `tabs`, and `drawer` types.
 * Renders all screens at once and injects a `focused` prop based on navigation state.
 *
 * Useful in tests to simulate navigation behavior with a lightweight navigator.
 */
const TestNavigatorImpl = ({ type, ...rest }: TestNavigatorImplProps): React.ReactElement => {
  const router = (
    type === 'drawer' ? DrawerRouter : type === 'stack' ? StackRouter : TabRouter
  ) as RouterFactory<AnyNavigationState, ExtractedAction, AnyRouterOptions>;

  const {
    state,
    descriptors,
    NavigationContent: Content,
  } = useNavigationBuilder<
    AnyNavigationState,
    AnyRouterOptions,
    AnyActionHelpers,
    TestScreenOptions,
    NavigationHelpers<ParamListBase>
  >(router, rest);

  return <Content>{state.routes.map((route) => descriptors[route.key]?.render())}</Content>;
};

/**
 * Creates a navigator factory for the TestNavigator.
 * This wraps the custom `TestNavigator` using `createNavigatorFactory`,
 * enabling use of `Navigator` and `Screen` components just like built-in navigators.
 *
 * Use this when integrating `TestNavigator` into `NavigationContainer`.
 */
const createTestNavigatorFactory = createNavigatorFactory(TestNavigatorImpl);

type LayoutProps = {
  name: string;
  focused: boolean;
  navigation: NavigationHelpers<ParamListBase>;
  children: React.ReactNode;
};

const CustomLayout = ({ name, focused, navigation, children }: LayoutProps): React.ReactElement => {
  useEffect(() => {
    const state = navigation.getState();
    const currentRoute = state.routes[state.index];

    if (focused && currentRoute?.name !== name) {
      navigation.navigate(name);
    } else if (!focused && currentRoute?.name !== 'Dummy') {
      navigation.navigate('Dummy');
    }
  }, [focused, name, navigation]);

  return <>{children}</>;
};

const TestLayout = ({ name, focused, ...props }: LayoutProps): React.ReactElement => {
  return <CustomLayout {...props} name={name} focused={focused} />;
};

export const TestNavigator = ({
  children,
  type = 'stack',
  name = 'TestScreen',
  focused = true,
  screenParams,
  dummyName = 'Dummy',
}: {
  children: React.ReactNode;
  type?: NavigatorType;
  name?: string;
  focused?: boolean;
  screenParams?: object;
  dummyName?: string;
}) => {
  const Test = createTestNavigatorFactory();

  return (
    <NavigationContainer>
      <Test.Navigator
        key={type}
        type={type}
        initialRouteName={focused ? name : dummyName}
        layout={
          /* eslint-disable-next-line react/no-unstable-nested-components */
          (props: Omit<LayoutProps, 'name' | 'focused'>) => (
            <TestLayout {...props} name={name} focused={focused} />
          )
        }
      >
        <Test.Screen name={name} getId={() => name} initialParams={screenParams}>
          {() => children}
        </Test.Screen>
        <Test.Screen name={dummyName} getId={() => dummyName}>
          {() => <Text testID="dummy-screen">Dummy</Text>}
        </Test.Screen>
      </Test.Navigator>
    </NavigationContainer>
  );
};
