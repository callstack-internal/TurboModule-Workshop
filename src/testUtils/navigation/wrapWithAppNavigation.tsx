import React from 'react';

import {
  InitialState,
  NavigationContainer,
  NavigationContainerRef,
} from '@react-navigation/native';
import DrawerNavigator from '../../navigation/navigators/DrawerNavigator.tsx';

type WrapWithAppNavigationProps = {
  initialState?: InitialState;
  navigationRef?: React.RefObject<NavigationContainerRef<any>>;
};

/**
 * Wraps the full app navigator (`DrawerNavigator`) in a `NavigationContainer` for testing.
 *
 * This is the preferred utility when writing tests for **real screen components or components that navigate** that are part of
 * the actual app navigation tree. It gives you a full, realistic navigation environment without mocks,
 * including nested navigators, screen options, transitions, etc.
 *
 * Unlike `TestNavigator`, this uses your actual app's navigator configuration, making it ideal for testing:
 * - Navigation behavior (e.g., `navigate`, `goBack`, deep linking)
 * - Screen lifecycle hooks (e.g., `useFocusEffect`, `useNavigation`)
 * - Integration-level rendering of screen flows
 *
 * ⚠️ Note: This should be used when testing **screen components**.
 * If you're only testing a non-screen component (e.g. custom hook, modal, header)
 * that requires navigation context, consider using `TestNavigator` or `createTestNavigator` instead for a minimal setup.
 *
 */
export const wrapWithAppNavigation = ({
  initialState,
  navigationRef,
}: WrapWithAppNavigationProps) => {
  return (
    <NavigationContainer initialState={initialState} ref={navigationRef}>
      <DrawerNavigator />
    </NavigationContainer>
  );
};
