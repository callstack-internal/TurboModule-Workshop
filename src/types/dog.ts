export interface Perro {
  nombre: string;
  raza: string;
  foto: string;
  edad?: number;
}

export interface Chip {
  chipId: string;
  fechaRegistro: string;
}

export interface NoAdoptable {
  motivo: 'enfermo' | 'muy joven';
  fechaDisponible: string;
}

export type PerroConChip = Perro & Chip;

export type PerroConEstado = Perro | (Perro & NoAdoptable);