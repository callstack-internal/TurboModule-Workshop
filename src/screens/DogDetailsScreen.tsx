import React from 'react';
import {
  View,
  Text,
  Image,
  StyleSheet,
  ScrollView,
  TouchableOpacity,
  Alert,
  SafeAreaView,
} from 'react-native';
import { useRoute } from '@react-navigation/native';
import { useAdoption } from '../context/AdoptionContext';
import { PerroConEstado } from '../types/dog';
import { esAdoptable } from '../utils/dogUtils';

function DogDetailsScreen() {
  const route = useRoute<{
    key: string;
    name: string;
    params: { dog: PerroConEstado };
  }>();
  const { addAdoptedDog, isAdopted } = useAdoption();

  // Check if route.params exists and has a dog property
  if (!route.params || !route.params.dog) {
    console.error('No dog data in route params');
    return (
      <SafeAreaView style={styles.container}>
        <View style={styles.errorContainer}>
          <Text style={styles.errorText}>Error: No se pudo cargar la información del perro</Text>
        </View>
      </SafeAreaView>
    );
  }

  const { dog } = route.params;

  const adoptable = esAdoptable(dog);
  const alreadyAdopted = isAdopted(dog.nombre);

  // Format dog age text
  const ageText = !dog.edad ? 'Edad desconocida' : dog.edad === 1 ? '1 año' : `${dog.edad} años`;

  // Handle adoption button press
  const handleAdopt = () => {
    if (!adoptable) {
      Alert.alert(
        'No disponible para adopción',
        'Este perro no está disponible para adopción en este momento.',
        [{ text: 'Entendido' }],
      );
      return;
    }

    if (alreadyAdopted) {
      Alert.alert('Ya adoptado', 'Ya has adoptado a este perro.', [{ text: 'Entendido' }]);
      return;
    }

    Alert.alert('Confirmar adopción', `¿Estás seguro de que quieres adoptar a ${dog.nombre}?`, [
      { text: 'Cancelar', style: 'cancel' },
      {
        text: 'Adoptar',
        onPress: () => {
          addAdoptedDog(dog);
          Alert.alert(
            '¡Felicidades!',
            `Has adoptado a ${dog.nombre}. Puedes ver a tus perros adoptados en la sección "Mis Perros".`,
            [{ text: 'OK' }],
          );
        },
      },
    ]);
  };

  if (!dog.foto) {
    return (
      <View style={styles.emptyContainer}>
        <Text style={styles.emptyText}>No se pudo cargar la información del perro</Text>
      </View>
    );
  }

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        <Image source={{ uri: dog.foto }} style={styles.image} resizeMode="cover" />

        <View style={styles.infoContainer}>
          <Text style={styles.name}>{dog.nombre}</Text>
          <Text style={styles.breed}>{dog.raza}</Text>
          <Text style={styles.age}>{ageText}</Text>

          {dog.distancia != null && (
            <View style={styles.distanceContainer}>
              <Text style={styles.distanceText}>A {dog.distancia} km de tu ubicación</Text>
            </View>
          )}

          <View style={styles.divider} />

          <View style={styles.statusContainer}>
            <View
              style={[
                styles.statusBadge,
                adoptable ? styles.adoptableBadge : styles.notAdoptableBadge,
              ]}
            >
              <Text style={styles.statusText}>
                {adoptable ? 'Disponible para adopción' : 'No adoptable aún'}
              </Text>
            </View>

            {!adoptable && 'motivo' in dog && (
              <Text style={styles.reasonText}>
                Motivo: {dog.motivo === 'enfermo' ? 'Está enfermo' : 'Es muy joven'}
              </Text>
            )}

            {!adoptable && 'fechaDisponible' in dog && (
              <Text style={styles.availableText}>
                Disponible a partir de: {dog.fechaDisponible}
              </Text>
            )}
          </View>

          <View style={styles.divider} />

          <Text style={styles.sectionTitle}>Sobre {dog.nombre}</Text>
          <Text style={styles.description}>
            {dog.nombre} es un perro de raza {dog.raza?.toLowerCase()}
            {dog.edad ? ` de ${dog.edad} ${dog.edad === 1 ? 'año' : 'años'}` : ''}. Es un compañero
            leal y cariñoso que busca un hogar donde pueda recibir todo el amor que merece.
          </Text>

          <TouchableOpacity
            style={[
              styles.adoptButton,
              !adoptable || alreadyAdopted ? styles.disabledButton : null,
            ]}
            onPress={handleAdopt}
            disabled={!adoptable || alreadyAdopted}
          >
            <Text style={styles.adoptButtonText}>
              {alreadyAdopted
                ? 'Ya has adoptado a este perro'
                : adoptable
                ? 'Adoptar'
                : 'No disponible para adopción'}
            </Text>
          </TouchableOpacity>
        </View>
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#ffffff',
  },
  image: {
    width: '100%',
    height: 300,
  },
  infoContainer: {
    padding: 16,
  },
  name: {
    fontSize: 28,
    fontWeight: 'bold',
    color: '#212529',
    marginBottom: 4,
  },
  breed: {
    fontSize: 18,
    color: '#495057',
    marginBottom: 4,
    textTransform: 'capitalize',
  },
  age: {
    fontSize: 16,
    color: '#6c757d',
    marginBottom: 8,
  },
  divider: {
    height: 1,
    backgroundColor: '#e9ecef',
    marginVertical: 16,
  },
  statusContainer: {
    marginTop: 4,
  },
  statusBadge: {
    paddingVertical: 6,
    paddingHorizontal: 12,
    borderRadius: 20,
    alignSelf: 'flex-start',
    marginBottom: 8,
  },
  adoptableBadge: {
    backgroundColor: '#d4edda',
  },
  notAdoptableBadge: {
    backgroundColor: '#f8d7da',
  },
  statusText: {
    fontSize: 14,
    fontWeight: '500',
    color: '#212529',
  },
  reasonText: {
    fontSize: 14,
    color: '#dc3545',
    marginTop: 4,
    fontStyle: 'italic',
  },
  availableText: {
    fontSize: 14,
    color: '#0066cc',
    marginTop: 4,
  },
  distanceContainer: {
    marginTop: 4,
    flexDirection: 'row',
    alignItems: 'center',
    flexWrap: 'wrap',
  },
  distanceText: {
    fontSize: 14,
    color: '#0066cc',
    fontWeight: '500',
  },
  sectionTitle: {
    fontSize: 20,
    fontWeight: 'bold',
    color: '#212529',
    marginBottom: 8,
  },
  description: {
    fontSize: 16,
    color: '#495057',
    lineHeight: 24,
    marginBottom: 16,
  },
  adoptButton: {
    backgroundColor: '#28a745',
    paddingVertical: 12,
    paddingHorizontal: 16,
    borderRadius: 8,
    alignItems: 'center',
    marginTop: 16,
  },
  disabledButton: {
    backgroundColor: '#6c757d',
  },
  adoptButtonText: {
    color: '#ffffff',
    fontWeight: 'bold',
    fontSize: 16,
  },
  errorContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 24,
  },
  errorText: {
    fontSize: 16,
    color: '#dc3545',
    textAlign: 'center',
  },
  emptyContainer: {
    padding: 24,
    alignItems: 'center',
  },
  emptyText: {
    fontSize: 16,
    color: '#6c757d',
    textAlign: 'center',
  },
});

export default DogDetailsScreen;
