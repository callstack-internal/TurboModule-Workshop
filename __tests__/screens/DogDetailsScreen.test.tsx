import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import { Alert } from 'react-native';
import { Screens } from '../../src/navigation/constants.ts';
import { wrapWithAppNavigation } from '../../src/testUtils/navigation/wrapWithAppNavigation.tsx';
import { TestNavigator } from '../../src/testUtils/navigation/TestNavigator.tsx';
import DogDetailsScreen from '../../src/screens/DogDetailsScreen';
import { PerroConEstado } from '../../src/types/dog';
import { wrapWithAdoptionContext } from '../../src/testUtils/context/wrapWithAdoptionContext.tsx';
import AsyncStorage from '@react-native-async-storage/async-storage';

// Sample adoptable dog for initialState
const initialDogDefault: PerroConEstado = {
  nombre: 'Max',
  raza: 'labrador',
  foto: 'https://example.com/dog.jpg',
  edad: 3,
  distancia: 2.5,
};

export const renderWithDogDetails = (initialDog: PerroConEstado | {} = initialDogDefault) => {
  const initialState = {
    routes: [
      {
        name: Screens.Main.name,
        state: {
          routes: [
            {
              name: Screens.DogDetails.name,
              params: { dog: initialDog },
            },
          ],
        },
      },
    ],
  };

  return render(wrapWithAdoptionContext(wrapWithAppNavigation({ initialState })));
};

describe('DogDetailsScreen', () => {
  // Sample adoptable dog
  const adoptableDog: PerroConEstado = {
    nombre: 'Max',
    raza: 'labrador',
    foto: 'https://example.com/dog.jpg',
    edad: 3,
    distancia: 2.5,
  };

  // Sample non-adoptable dog
  const nonAdoptableDog: PerroConEstado = {
    nombre: 'Luna',
    raza: 'golden retriever',
    foto: 'https://example.com/dog2.jpg',
    edad: 1,
    motivo: 'muy joven',
    fechaDisponible: '2023-12-31',
  };

  beforeEach(async () => {
    await AsyncStorage.clear(); // Clear before each test
    jest.clearAllMocks();
  });

  afterAll(async () => {
    await AsyncStorage.clear(); // Clear after all test
  });

  test('renders adoptable dog details correctly', async () => {
    const { getByText } = renderWithDogDetails();

    await waitFor(() => {
      expect(getByText('Max')).toBeTruthy();
      expect(getByText('labrador')).toBeTruthy();
      expect(getByText('3 años')).toBeTruthy();
      expect(getByText('A 2.5 km de tu ubicación')).toBeTruthy();
      expect(getByText('Disponible para adopción')).toBeTruthy();
      expect(getByText('Adoptar')).toBeTruthy();
    });
  });

  test('renders non-adoptable dog details correctly', () => {
    const { getByText } = renderWithDogDetails(nonAdoptableDog);

    // Check if basic information is displayed
    expect(getByText('Luna')).toBeTruthy();
    expect(getByText('golden retriever')).toBeTruthy();
    expect(getByText('1 año')).toBeTruthy();

    // Check if the correct status is displayed
    expect(getByText('No adoptable aún')).toBeTruthy();
    expect(getByText('Motivo: Es muy joven')).toBeTruthy();
    expect(getByText('Disponible a partir de: 2023-12-31')).toBeTruthy();

    // Check if the adopt button is disabled
    expect(getByText('No disponible para adopción')).toBeTruthy();
  });

  test('renders error when no dog data is provided', () => {
    const { getByText } = renderWithDogDetails({});

    // Check if error message is displayed
    expect(getByText('No se pudo cargar la información del perro')).toBeTruthy();
  });

  test('shows already adopted message when dog is already adopted', async () => {
    // Manually persist the dog into AsyncStorage before rendering
    await AsyncStorage.setItem(
      'adoptedDogs',
      JSON.stringify([adoptableDog]), // simulate previously adopted
    );
    const { getByText } = renderWithDogDetails();

    await waitFor(() => {
      // Check if the already adopted message is displayed
      expect(getByText('Ya has adoptado a este perro')).toBeTruthy();
    });
  });

  test('adopts a dog when adopt button is pressed', () => {
    const { getByText } = renderWithDogDetails();

    // Press the adopt button
    fireEvent.press(getByText('Adoptar'));

    // Check if confirmation alert was shown
    expect(Alert.alert).toHaveBeenCalledWith(
      'Confirmar adopción',
      '¿Estás seguro de que quieres adoptar a Max?',
      expect.arrayContaining([
        expect.objectContaining({ text: 'Cancelar' }),
        expect.objectContaining({ text: 'Adoptar' }),
      ]),
    );

    // Simulate pressing the Adoptar button in the alert
    const adoptHandler = (Alert.alert as jest.Mock).mock.calls[0][2][1].onPress;

    act(() => adoptHandler());

    // Check if success alert was shown
    expect(Alert.alert).toHaveBeenCalledWith(
      '¡Felicidades!',
      'Has adoptado a Max. Puedes ver a tus perros adoptados en la sección "Mis Perros".',
      expect.arrayContaining([expect.objectContaining({ text: 'OK' })]),
    );
  });

  test('shows message when trying to adopt a non-adoptable dog', () => {
    const { getByText } = renderWithDogDetails(nonAdoptableDog);

    expect(getByText('No disponible para adopción')).toBeTruthy();
  });

  test('matches snapshot', async () => {
    const { toJSON } = render(
      <TestNavigator name="DogDetails" screenParams={{ dog: adoptableDog }}>
        <DogDetailsScreen />
      </TestNavigator>,
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
