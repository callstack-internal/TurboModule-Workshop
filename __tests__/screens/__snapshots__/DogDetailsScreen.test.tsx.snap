// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DogDetailsScreen matches snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <Image
        resizeMode="cover"
        source={
          {
            "uri": "https://example.com/dog.jpg",
          }
        }
        style={
          {
            "height": 300,
            "width": "100%",
          }
        }
      />
      <View
        style={
          {
            "padding": 16,
          }
        }
      >
        <Text
          style={
            {
              "color": "#212529",
              "fontSize": 28,
              "fontWeight": "bold",
              "marginBottom": 4,
            }
          }
        >
          Max
        </Text>
        <Text
          style={
            {
              "color": "#495057",
              "fontSize": 18,
              "marginBottom": 4,
              "textTransform": "capitalize",
            }
          }
        >
          labrador
        </Text>
        <Text
          style={
            {
              "color": "#6c757d",
              "fontSize": 16,
              "marginBottom": 8,
            }
          }
        >
          3 años
        </Text>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "marginTop": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "#0066cc",
                "fontSize": 14,
                "fontWeight": "500",
              }
            }
          >
            A 
            2.5
             km de tu ubicación
          </Text>
        </View>
        <View
          style={
            {
              "backgroundColor": "#e9ecef",
              "height": 1,
              "marginVertical": 16,
            }
          }
        />
        <View
          style={
            {
              "marginTop": 4,
            }
          }
        >
          <View
            style={
              [
                {
                  "alignSelf": "flex-start",
                  "borderRadius": 20,
                  "marginBottom": 8,
                  "paddingHorizontal": 12,
                  "paddingVertical": 6,
                },
                {
                  "backgroundColor": "#d4edda",
                },
              ]
            }
          >
            <Text
              style={
                {
                  "color": "#212529",
                  "fontSize": 14,
                  "fontWeight": "500",
                }
              }
            >
              Disponible para adopción
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "backgroundColor": "#e9ecef",
              "height": 1,
              "marginVertical": 16,
            }
          }
        />
        <Text
          style={
            {
              "color": "#212529",
              "fontSize": 20,
              "fontWeight": "bold",
              "marginBottom": 8,
            }
          }
        >
          Sobre 
          Max
        </Text>
        <Text
          style={
            {
              "color": "#495057",
              "fontSize": 16,
              "lineHeight": 24,
              "marginBottom": 16,
            }
          }
        >
          Max
           es un perro de raza 
          labrador
           de 3 años
          . Es un compañero leal y cariñoso que busca un hogar donde pueda recibir todo el amor que merece.
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#28a745",
              "borderRadius": 8,
              "marginTop": 16,
              "opacity": 1,
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "#ffffff",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Adoptar
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
